packages/pluggableWidgets/bar-chart-native/src/utils/SeriesLoader.ts

```diff
@@ -0,0 +1,230 @@
+            dataPoints: dataPointsExtraction.dataPoints,
+            xFormatter: dataPointsExtraction.xFormatter,
+            yFormatter: dataPointsExtraction.yFormatter,
+            name: itemGroup.dynamicSeriesNameValue,
+            customBarStyle: itemGroup.dynamicCustomBarStyleValue
+        });
+    }
+
+    return loadedSeries;
+}
+
+function groupDataSourceItems(series: BarSeriesType): DataSourceItemGroup[] | null {
+    const { dynamicDataSource, dynamicSeriesName, dynamicCustomBarStyle, groupByAttribute, dataSet } = series;
+
+    if (dataSet !== "dynamic") {
+        throw Error("Expected series to be dynamic");
+    }
+
+    const dataSource = ensure(dynamicDataSource);
+
+    if (!dataSource.items) {
+        return null;
+    }
+
+    const dataSourceItemGroupsResult: DataSourceItemGroup[] = [];
+
+    for (const item of dataSource.items) {
+        const groupByAttributeValue = ensure(groupByAttribute)(item);
+
+        if (groupByAttributeValue.value === undefined) {
+            return null;
+        }
+
+        const group = dataSourceItemGroupsResult.find(group => {
+            if (groupByAttributeValue.value instanceof Date && group.groupByAttributeValue instanceof Date) {
+                return group.groupByAttributeValue.getTime() === groupByAttributeValue.value.getTime();
+            } else if (groupByAttributeValue.value instanceof Big && group.groupByAttributeValue instanceof Big) {
+                return group.groupByAttributeValue.eq(groupByAttributeValue.value);
+            }
+
+            return group.groupByAttributeValue === groupByAttributeValue.value;
+        });
+
+        if (group) {
+            group.items.push(item);
+        } else {
+            const newDataSourceItemGroup: DataSourceItemGroup = {
+                groupByAttributeValue: groupByAttributeValue.value,
+                items: [item]
+            };
+
+            if (dynamicSeriesName) {
+                const dynamicSeriesNameValue = dynamicSeriesName(item);
+
+                if (dynamicSeriesNameValue.value === undefined) {
+                    return null;
+                }
+
+                newDataSourceItemGroup.dynamicSeriesNameValue = dynamicSeriesNameValue.value;
+            }
+
+            if (dynamicCustomBarStyle) {
+                const dynamicCustomBarStyleValue = dynamicCustomBarStyle(item);
+
+                if (dynamicCustomBarStyleValue.value === undefined) {
+                    return null;
+                }
+
+                newDataSourceItemGroup.dynamicCustomBarStyleValue = dynamicCustomBarStyleValue.value;
+            }
+
+            dataSourceItemGroupsResult.push(newDataSourceItemGroup);
+        }
+    }
+
+    return dataSourceItemGroupsResult;
+}
+
+function extractDataPoints(series: BarSeriesType, dataSourceItems?: ObjectItem[]): DataPointsExtraction | null {
+    if (!dataSourceItems) {
+        // todo: why is this check here? dataSourceItems are only passed for dynamic items.
```

Comment:
Correct, we can leave out this check.

Reply:
Removed the ternary, now the variable is always set to the static data source `ListValue`.